fuzz/src/main/java/edu/berkeley/cs/jqf/fuzz/util/FastNonCollidingCoverage.java

```diff
@@ -0,0 +1,234 @@
+
+        int ret = 1;
+
+        while ((num >>= 1) != 0)
+            ret <<= 1;
+
+        return ret;
+    }
+
+    /** Populates the HOB cache. */
+    static {
+        for (int i = 0; i < HOB_CACHE.length; i++) {
+            HOB_CACHE[i] = computeHob(i);
+        }
+    }
+
+    /** Returns the highest order bit (perhaps using the cache) */
+    private static int hob(int num) {
+        if (num < HOB_CACHE.length) {
+            return HOB_CACHE[num];
+        } else {
+            return computeHob(num);
+        }
+    }
+
+
+    /**
+     * Updates this coverage with bits from the parameter.
+     *
+     * @param that the run coverage whose bits to OR
+     *
+     * @return <code>true</code> iff <code>that</code> is not a subset
+     *         of <code>this</code>, causing <code>this</code> to change.
+     */
+    public boolean updateBits(ICoverage that) {
+        boolean changed = false;
+        synchronized (this.counter){
+            synchronized (that.getCounter()){
+                FastNonCollidingCounter thatCounter = (FastNonCollidingCounter) that.getCounter();
+                Iterator<IntIntPair> thatIter = thatCounter.counts.keyValuesView().iterator();
+
+                while(thatIter.hasNext()){
+                    IntIntPair coverageEntry = thatIter.next();
+                    int before = this.counter.counts.get(coverageEntry.getOne());
+                    int after = before | hob(coverageEntry.getTwo());
+                    if(after != before){
+                        this.counter.counts.put(coverageEntry.getOne(), after);
+                        changed = true;
+                    }
+                    if(before == 0){
+                        this.counter.nonZeroKeys.add(coverageEntry.getOne());
+                    }
+                }
+            }
+        }
+        return changed;
+    }
+
+    /** Returns a hash code of the edge counts in the coverage map. */
+    @Override
+    public int hashCode() {
+        return counter.counts.hashCode();
+    }
+
+    /**
+     * Returns a hash code of the list of edges that have been covered at least once.
+     *
+     * @return a hash of non-zero entries
+     */
+    public int nonZeroHashCode() {
+        return counter.getNonZeroIndices().hashCode();
+    }
+
+    @Override
+    public Counter getCounter() {
+        return this.counter;
+    }
+
+    /**
+     * @return a string representing the counter
+     */
+    @Override
+    public String toString() {
+        StringBuffer sb = new StringBuffer();
+        sb.append("Coverage counts: \n");
+        for (int i = 0; i < counter.counts.size(); i++) {
+            if (counter.counts.get(i) == 0) {
+                continue;
+            }
+            sb.append(i);
+            sb.append("->");
+            sb.append(counter.counts.get(i));
+            sb.append('\n');
+        }
+        return sb.toString();
+    }
+
+    @Override
+    public void logCoverage(int iid, int arm) {
```

Comment:
Do we need a more advanced way to log coverage instead of adding them together to avoid the collision? For example, given the following function

```
    public static int test(int a) {
        int b = 0;
        if (a > 0) {
            b += 1;  // iid = 1, arm = 1
        } else {
            b -= 1; // iid = 1, arm = 0
        }
        if (a / 2 == 0) {
            b += 1;  // iid = 2, arm = 1
        } else {
            b -= 1; // iid = 2, arm = 0
        }
        return b;
    }
```

input `a=11` leads to coverage `{2: 2}`. This isn't ideal since we want to distinguish two branches.

Shall we either change here or the `FastCoverageMethodAdapter` to preallocate iids for different branch?

Reply:
> Shall we either change here or the FastCoverageMethodAdapter to preallocate iids for different branch?

There is an undocumented contract that callers should create enough IIDs to allocate the space. We should probably add this in documentation. AFAIK, the FastCoverageMethodAdapter allocates enough probes to avoid collisions for branches and for switch statements.

Reply:
Sorry, there was a typo in the example, the condition of the second branch should be `a % 2 == 0`. 

Yes, `FastCoverageMethodAdapter` allocates enough IIDs but different execution points will be mapped to the same location. 

In the above example, `BRANCH1T` and `BRANCH2F` will have the same index `2` in `FastNonCollidingCounter` because the index of `BRANCH1T` is `iid + arm = 1 + 1`, and the index of `BRANCH2F` is `iid + arm = 2 + 0`.  

What I'm suggesting is to allocate IIDs for each branch: for example, if the first branch has two arms, then we reserve the IID(2) for the true arm of the first branch. Then the IID for the second branch should be 3 instead of 2 and this will avoid collisions.

Reply:
I am confused. We are looking at the same code: [here](https://github.com/jon-bell/JQF/blob/fast-collision-free-coverage-clean/instrument/src/main/java/janala/instrument/FastCoverageMethodAdapter.java#L80), right? The correct IIDs and arms as per this code (as far as I can tell) are:

```java
    public static int test(int a) {
        int b = 0;
        if (a > 0) {
            b += 1;  // BRANCH1T,  iid = 1, arm = 1
        } else {
            b -= 1; // BRANCH1F, iid = 1, arm = 0
        }
        if (a % 2 == 0) {
            b += 1;  // BRANCH2T, iid = 3, arm = 1
        } else {
            b -= 1; // BRANCH2F, iid = 3, arm = 0
        }
        return b;
    }
```

Then, the coverage probes evaluated at runtime (adding IID and arm) will give us: `(1, 2, 3, 4)` - all unique.

Reply:
Ah, I see the issue, the https://github.com/jon-bell/JQF/blob/fast-collision-free-coverage-clean/instrument/src/main/java/janala/instrument/FastCoverageMethodAdapter.java#L81 should be 

```
instrumentationState.incAndGetFastCoverageId();
```

That's the reason I'm getting duplicate IIDs.

Reply:
ahhhhhh thank you for catching this - sorry! I had lifted this patch off of another branch of JQF that I have where I had simply removed all of the existing coverage instrumentation - and in that branch I used the existing `incAndGetId`. When I prepared the PR I copy and pasted from that and thought that I had gotten it all lined up. I will run another 24 hour campaign with this fix; I am curious to see how it impacts the overall performance.

Reply:
Nice catch, @aoli-al! That's quite nuanced. And thanks for the quick fix, Jon. I'm impressed that the results were already so good even with these accidental collisions.